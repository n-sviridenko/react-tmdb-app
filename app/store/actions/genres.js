import { createAction } from 'redux-actions';

export const LIST_REQUEST = 'app/genres/LIST_REQUEST';
export const LIST_SUCCESS = 'app/genres/LIST_SUCCESS';
export const LIST_FAILURE = 'app/genres/LIST_FAILURE';

export const listRequest = createAction(LIST_REQUEST);
export const listSuccess = createAction(LIST_SUCCESS);
export const listFailure = createAction(LIST_FAILURE);
