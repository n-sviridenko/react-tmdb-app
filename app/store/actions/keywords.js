import { createAction } from 'redux-actions';

export const SEARCH_REQUEST = 'app/keywords/SEARCH_REQUEST';
export const SEARCH_SUCCESS = 'app/keywords/SEARCH_SUCCESS';
export const SEARCH_FAILURE = 'app/keywords/SEARCH_FAILURE';

export const searchRequest = createAction(SEARCH_REQUEST, null, (query) => ({ key: query }));
export const searchSuccess = createAction(SEARCH_SUCCESS, null, (res, key) => ({ key }));
export const searchFailure = createAction(SEARCH_FAILURE, null, (err, key) => ({ key }));

export const SHOW_REQUEST = 'app/keywords/SHOW_REQUEST';
export const SHOW_SUCCESS = 'app/keywords/SHOW_SUCCESS';
export const SHOW_FAILURE = 'app/keywords/SHOW_FAILURE';

export const showRequest = createAction(SHOW_REQUEST, null, (id) => ({ key: id }));
export const showSuccess = createAction(SHOW_SUCCESS, null, (res, key) => ({ key }));
export const showFailure = createAction(SHOW_FAILURE, null, (err, key) => ({ key }));
