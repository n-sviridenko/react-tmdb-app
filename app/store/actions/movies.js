import { createAction } from 'redux-actions';

export const DISCOVER_REQUEST = 'app/movies/DISCOVER_REQUEST';
export const DISCOVER_SUCCESS = 'app/movies/DISCOVER_SUCCESS';
export const DISCOVER_FAILURE = 'app/movies/DISCOVER_FAILURE';

export const discoverRequest = createAction(DISCOVER_REQUEST);
export const discoverSuccess = createAction(DISCOVER_SUCCESS);
export const discoverFailure = createAction(DISCOVER_FAILURE);
