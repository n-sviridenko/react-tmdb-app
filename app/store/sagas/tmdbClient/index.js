export sendListRequest from './sendListRequest';
export sendRequest from './sendRequest';
