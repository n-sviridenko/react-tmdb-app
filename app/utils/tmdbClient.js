import get from 'lodash/get';
import isString from 'lodash/isString';
import theMovieDb from 'themoviedb-javascript-library';

export function sendRequest(methodName, ...args) {
  const method = get(theMovieDb, methodName);

  return new Promise((resolve, reject) => {
    // Ensure that the response was decoded from JSON
    // TODO: use resolve when https://github.com/cavestri/themoviedb-javascript-library/issues/36 will be resolved
    const finalResolve = (data) => {
      let finalData = data;

      if (isString(finalData)) {
        try {
          finalData = JSON.parse(data);
        } catch (e) {} // eslint-disable-line no-empty
      }

      resolve(finalData);
    };

    method(...args, finalResolve, reject);
  });
}
