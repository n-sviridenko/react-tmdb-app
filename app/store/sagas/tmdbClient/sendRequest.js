import { call, put } from 'redux-saga/effects';
import { fromJS } from 'immutable';
import { normalize } from 'normalizr';

import { sendRequest as baseSendRequest } from 'utils/tmdbClient';
import { merge } from 'store/actions/entities';

function* sendRequest(methodName, responseSchema, ...args) {
  const response = yield call(baseSendRequest, methodName, ...args);

  if (responseSchema === null) {
    return fromJS(response);
  }

  const { result, entities } = normalize(response, responseSchema);

  yield put(merge(fromJS(entities)));

  return fromJS(result);
}

export default sendRequest;
