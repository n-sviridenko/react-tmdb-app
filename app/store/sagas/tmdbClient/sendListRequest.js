import { call } from 'redux-saga/effects';
import { Map } from 'immutable';

import * as schemas from 'store/schemas';
import sendRequest from './sendRequest';

function* sendListRequest(methodName, itemSchema, ...args) {
  const res = yield call(sendRequest, methodName, schemas.makeList(itemSchema), ...args);

  return Map({
    count: res.get('total_results'),
    pageCount: res.get('total_pages'),
    page: res.get('page'),
    items: res.get('results'),
  });
}

export default sendListRequest;
