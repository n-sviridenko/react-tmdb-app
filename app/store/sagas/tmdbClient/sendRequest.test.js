/* eslint-disable redux-saga/yield-effects */
import { call, put } from 'redux-saga/effects';
import { fromJS } from 'immutable';
import { schema } from 'normalizr';

import { sendRequest as baseSendRequest } from 'utils/tmdbClient';
import { merge } from 'store/actions/entities';
import sendRequest from './sendRequest';

describe('tmdbClient sendRequest saga', () => {
  it('should send a request', () => {
    const sendRequestTask = sendRequest('discover.getMovies', null, { foo: 'bar' });
    const callEffect = call(baseSendRequest, 'discover.getMovies', { foo: 'bar' });
    expect(sendRequestTask.next().value).toEqual(callEffect);
  });

  it('should return an immutable response', () => {
    const sendRequestTask = sendRequest('discover.getMovies', null, { foo: 'bar' });
    expect(sendRequestTask.next().value).toMatchSnapshot();
    const returnDescriptor = sendRequestTask.next({});
    expect(returnDescriptor.value).toEqual(fromJS({}));
    expect(returnDescriptor.done).toBe(true);
  });

  it('should return normalize response if schema is specified', () => {
    const responseSchema = new schema.Entity('movie');
    const sendRequestTask = sendRequest('discover.getMovie', responseSchema, { id: 1 });
    expect(sendRequestTask.next().value).toMatchSnapshot();
    const movie = { id: 1, title: 'Sam' };
    const entities = fromJS({ movie: { 1: movie } });
    const mergeDescriptor = sendRequestTask.next(movie);
    expect(mergeDescriptor.value).toEqual(put(merge(entities)));
    const returnDescriptor = sendRequestTask.next();
    expect(returnDescriptor.value).toBe(1);
    expect(returnDescriptor.done).toBe(true);
  });
});
