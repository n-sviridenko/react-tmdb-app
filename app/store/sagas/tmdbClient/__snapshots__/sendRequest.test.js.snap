exports[`sendRequest saga should return an immutable response 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "discover.getMovies",
      Object {
        "foo": "bar",
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`sendRequest saga should return normalize response if schema is specified 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "discover.getMovie",
      Object {
        "id": 1,
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
